lib/logstash/outputs/webhdfs.rb

```diff
@@ -0,0 +1,278 @@
+
+  # How many times should we retry. If retry_times is exceeded, an error will be logged and the event will be discarded.
+  config :retry_times, :validate => :number, :default => 5
+
+  # Compress output. One of ['none', 'snappy', 'gzip']
+  config :compression, :validate => ["none", "snappy", "gzip"], :default => "none"
+
+  # Set snappy chunksize. Only neccessary for stream format. Defaults to 32k. Max is 65536
+  # @see http://code.google.com/p/snappy/source/browse/trunk/framing_format.txt
+  config :snappy_bufsize, :validate => :number, :default => 32768
+
+  # Set snappy format. One of "stream", "file". Set to stream to be hive compatible.
+  config :snappy_format, :validate => ["stream", "file"], :default => "stream"
+
+  # Remove @timestamp field. Hive does not like a leading "@", but we need @timestamp for path calculation.
+  config :remove_at_timestamp, :validate => :boolean, :default => true
+
+  public
+
+  def register
+    begin
+      require 'webhdfs'
+    rescue LoadError
+      @logger.error("Module webhdfs could not be loaded.")
+      raise
+    end
+    if @compression == "gzip"
+      begin
+        require "zlib"
+      rescue LoadError
+        @logger.error("Gzip compression selected but zlib module could not be loaded.")
+        raise
+      end
+    elsif @compression == "snappy"
+      begin
+        require "snappy"
+      rescue LoadError
+        @logger.error("Snappy compression selected but snappy module could not be loaded.")
+        raise
+      end
+    end
+    @files = {}
+    @client = prepare_client(@host, @port, @user)
+    # Test client connection.
+    begin
+      @client.list('/')
+    rescue => e
+      @logger.error("Webhdfs check request failed. (namenode: #{@client.host}:#{@client.port}, Exception: #{e.message})")
+      raise
+    end
+    buffer_initialize(
+      :max_items => @flush_size,
+      :max_interval => @idle_flush_time,
+      :logger => @logger
+    )
+  end # def register
+
+  public
+  def receive(event)
+    return unless output?(event)
+    buffer_receive(event)
+  end # def receive
+
+  def prepare_client(host, port, username)
+    client = WebHDFS::Client.new(host, port, username)
+    client.httpfs_mode = @use_httpfs
+    client.open_timeout = @open_timeout
+    client.read_timeout = @read_timeout
+    client.retry_known_errors = @retry_known_errors
+    client.retry_interval = @retry_interval if @retry_interval
+    client.retry_times = @retry_times if @retry_times
+    client
+  end
+
+  def flush(events=nil, teardown=false)
+    return if not events
+    # Avoid creating a new string for newline every time
+    newline = "\n".freeze
+    output_files = Hash.new { |hash, key| hash[key] = "" }
+    events.collect do |event|
+      path = event.sprintf(@path)
+      if @remove_at_timestamp
+        event.remove("@timestamp")
+      end
+      if @message_format
+        event_as_string = event.sprintf(@message_format)
+      else
+        event_as_string = event.to_json
```

Comment:
To transform to json you should use the code provided in https://github.com/elastic/logstash/blob/master/lib/logstash/json.rb, that leverages jackson to be faster.